refactor(plugins):  BigNumber Time Comparison with existing time_offset API

Antonio-RiveroMartnez opened the pull request:
### SUMMARY
Previously we implemented the time comparison in BigNumber viz using some new custom properties in the queryObject to compute the offsets. We are moving away from that approach in order to be more consistent with the way Superset handles time comparison in other visualizations.

Some of the changes of this PR are:
- Use existing time_offset APIs instead of the instant comparison ones
- Create new shared section to include the controls for time comparison, some taken from Advanced Analytics
- Create a new control label to display the comparison ranges using existing time_range API and not a new one

There will be a follow up PR cleaning up the customizations from previous approach, that clean up is left out of this PR to ease the review process.


### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->

### TESTING INSTRUCTIONS
<!--- Required! What steps can be taken to manually verify the changes? -->

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [X] Required feature flags: FEATURE_CHART_PLUGINS_EXPERIMENTAL
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

Antonio-RiveroMartnez commented:
/testenv up

Antonio-RiveroMartnez commented:
/testenv up FEATURE_CHART_PLUGINS_EXPERIMENTAL=true

Antonio-RiveroMartnez commented:
/testenv up FEATURE_CHART_PLUGINS_EXPERIMENTAL=true

Antonio-RiveroMartnez committed:
BigNumber with Time Comparison:

- Use existig time_offset APIs instead of the instant comparison ones
- Create new shared section to include the controls for time comparison, some taken from Advanced Analytics
- Create a new control label to display the comparison ranges using existing time_range API and not a new one

Notes:
- Tooltip is no longer displaying comparison ranges in chart
- Time shift doesnt include Custom nor Inherit options
- Time grain is part of the controls but it might be removed

Antonio-RiveroMartnez committed:
BigNumber with Time Comparison:

- Use core method to render comparison date in tooltip

lilykuang committed:
implement start date offset controls

Antonio-RiveroMartnez committed:
Table with Time Comparison:

- Support textual columns in time_offsets comparisons
- Because textual columns have no time grain nor they use xaxis labels all the time, the time_grain and xaxis are not required for the comparison queries that use textual columns only

cherry picked from 5819e3a527c67c4357db0cef5e2e8b0d47624d37

kgabryje committed:
Fixes

cherry picked from 669d060ab1f94218449c609596c04f6c6f924e41

kgabryje committed:
Improve coverage

cherry picked from c1f73ab9611b14d466074739b788f06f05104684

lilykuang committed:
fix time shift

cherry picked from 6cacbe6519811d3d0d7907202f9d997336ce225f

lilykuang committed:
custom and inherit start date

cherry picked from d3d3b31453847244f7b17a58ded7d39ce4b5b514

Antonio-RiveroMartnez committed:
Table with Time Comparison:

- When calling the time range method ensure we pass an array so we can remove the array check and follow the method signature

Antonio-RiveroMartnez committed:
Big Number with Time Comparison:

- Remove time grain from time_comparison controls

Antonio-RiveroMartnez committed:
Table with Time Comparison:

- Support time comparison with no columns, meaning we want to compare just metrics. If so, we do an outer join

Antonio-RiveroMartnez committed:
BigNumber with Time Comparison:

- Do not pass a column because  it's not required anymore

Antonio-RiveroMartnez commented:
/testenv up FEATURE_CHART_PLUGINS_EXPERIMENTAL=true

lilykuang committed:
added getTimeOffset

Antonio-RiveroMartnez committed:
BigNumber with Time Comparison:

- Use the new getTimeOffset to render the correct time range in the tooltip
- Update the lock file with moment

Antonio-RiveroMartnez committed:
BigNumber with Time Comparison:

- Process data when custom or inherit offset is used
- Handle responses from time_range api that don't include the comparison ranges

Antonio-RiveroMartnez committed:
BigNumber with Time Comparison:

- Add tests for the new getTimeOffset method
- Hide label if no comparison is being made

Antonio-RiveroMartnez commented:
/testenv up FEATURE_CHART_PLUGINS_EXPERIMENTAL=true

lilykuang committed:
user parseDttmToMoment

eschutho commented on superset-frontend/packages/superset-ui-core/src/time-comparison/getTimeOffset.ts:
```diff
@@ -0,0 +1,101 @@
+
+export const parseDttmToMoment = (dttm: string): Moment => {
+  if (dttm === 'now') {
+    return moment().utc().startOf('second');
```
you should be able to do this with just vanilla javascript instead of Moment. We're currently trying to remove moment wherever possible.

lilykuang committed:
change parse dttm from moment to date

lilykuang committed:
update tests

Antonio-RiveroMartnez committed:
BigNumber with Time Comparison:

- Update lock file so we remove moment
- Stop displaying Calculation Type since this viz only uses Actual Values (postporocessing: [])
- Handle nullish dttm

Antonio-RiveroMartnez commented:
/testenv up FEATURE_CHART_PLUGINS_EXPERIMENTAL=true

Antonio-RiveroMartnez commented:
/testenv up FEATURE_CHART_PLUGINS_EXPERIMENTAL=true

Antonio-RiveroMartnez committed:
BigNumber with Time Comparison:

- Add new prop so we can show full list of choices or just a reduced amount based on the boolean value
- Update the copy so it better reflects and help the user

eschutho commented on superset-frontend/plugins/plugin-chart-echarts/src/BigNumber/BigNumberPeriodOverPeriod/controlPanel.ts:
```diff
@@ -180,14 +108,12 @@ const config: ControlPanelConfig = {
         ],
       ],
     },
+    sections.timeComparisonControls(false, false, false),
```
I'd suggest passing an object here instead of arguments so that we can tell which values are being set to false here for example.

Antonio-RiveroMartnez committed:
Time Comparison:

- Add migration to port from old time comparison params to new approach
- Add tests for migration file

michael-s-molina commented on superset/common/query_context_processor.py:
```diff
@@ -488,21 +511,17 @@ def processing_time_offsets(  # pylint: disable=too-many-locals,too-many-stateme
-                        )
+                if dataframe_utils.is_datetime_series(offset_metrics_df.get(index)):
+                    # modifies temporal column using offset
+                    offset_metrics_df[index] = offset_metrics_df[index] - DateOffset(
```
Hi @Antonio-RiveroMartnez. This logic to calculate the offsets is not correct and I'm fixing it in https://github.com/apache/superset/pull/28432. I think it would be prudent to wait for that fix and rebase your PR to see if it will affect this new feature in any way.

Antonio-RiveroMartnez replied:
Hey @michael-s-molina , thanks for the heads up! Ok, will do, it's very likely that your PR gets merged before this one so it should be ok for rebasing.

Antonio-RiveroMartnez committed:
Time Comparison:

- Fix delta calculation for shifts
- Add tests for new changes

eschutho reviewed (approved):
lgtm

michael-s-molina commented:
@Antonio-RiveroMartnez I added the `hold:testing` label to this PR. I'll wait for your rebase and will test the changes using Airbnb's production data, the same way I did for https://github.com/apache/superset/pull/28432 just to confirm that the time-shifts remain in sync with legacy charts.

michael-s-molina commented on superset-frontend/packages/superset-ui-core/test/time-comparison/getTimeOffset.test.ts:
```diff
@@ -0,0 +1,162 @@
+ */
+import { getTimeOffset, parseDttmToDate } from '@superset-ui/core';
+
+describe('parseDttmToDate', () => {
```
Can you avoid `describe` and just use `test` to follow this [best practice](https://github.com/apache/superset/wiki/Testing-Guidelines-and-Best-Practices#avoid-nesting-when-youre-testing)? We're slowly refactoring the tests to move in that direction. 

I noticed that this file name is `getTimeOffset.test.ts` but it's also testing `parseDttmToDate`. Can you rename the file to something more generic or breaking it into 2 test files?

Antonio-RiveroMartnez replied:
Done!

michael-s-molina commented on superset-frontend/packages/superset-ui-core/src/time-comparison/getTimeOffset.ts:
```diff
@@ -0,0 +1,121 @@
+import { ensureIsArray } from '../utils';
+
+const DAY_IN_MS = 24 * 60 * 60 * 1000;
+export const parseDttmToDate = (dttm: string): Date => {
```
Isn't `parseDttmToDate` similar to [parse_human_datetime](https://github.com/apache/superset/blob/821c7d7f2c430c4a4294883a66128ba98fd949c5/superset/utils/date_parser.py#L57)? If yes, is there a way we can reuse that so we don't have 2 implementations to maintain?

lilykuang replied:
Thank you for pointing this out. parseDttmToDate and parse_human_datetime are very similar. However, parse_human_datetime is in the backend, while parseDttmToDate is also for validating a [DatePicker component](https://github.com/apache/superset/pull/27718/files#diff-76b79c56cfbdda04728e5590bd1927760383d5c44c9d265a828da8804472a271R65-R81) in the frontend. This separation makes unifying them tricky. I'm open to any suggestions you might have.

michael-s-molina commented on superset-frontend/packages/superset-ui-core/test/time-comparison/fetchTimeRange.test.ts:
```diff
@@ -116,3 +117,54 @@ it('returns a formatted error message from response', async () => {
   });
 });
+
+it('fetchTimeRange with shift', async () => {
```
Same comment to use `test` instead of `it`.

Antonio-RiveroMartnez replied:
Done!

Antonio-RiveroMartnez committed:
Time Comparison:

- Fix lint error
- Fix tests

Antonio-RiveroMartnez committed:
Time Comparison:

- Pass a config object instead of raw bollean flags to the time comparison control so the intention is clear when using it
- Split getTimeOffset tests so parseDttmToDate has its own file
- use inlined test instead of blocks with describe and it

michael-s-molina commented:
Thanks for rebasing and addressing the comments @Antonio-RiveroMartnez. I'll execute the tests using Airbnb's production data.

Antonio-RiveroMartnez committed:
Time Comparison:

- Fix how we separate the time_grain scenario with join keys
- Simplify when we raise time grain required exception

Antonio-RiveroMartnez committed:
- Fix migration order
